### DISCUSSION TITLE
ISMAGS GSOC '24 Project

### DISCUSSION THREAD

**Discussion by elshorbagyx** [networkx/networkx#7359] [General]
Hi,

My name is Mohamed Rezk, a computer engineering student interested in graph algorithms. This year, I am interested in participating in Google Summer of Code, and I came across the ISMAGS code incorporation project. However, I realized that the project was offered last year as well. Is it still relevant? Moreover, I have some questions on my mind:

- What is the best way to start digging into this project?
- What kind of roadmap needs to be included in the proposal?
-  What criteria are needed to assess the Sandia Labs code?

I believe @dschult and @rossbar are the mentors. Any help would be appreciated.

Thanks for having me :)

**Comment by dschult** [comment]
Hello Mohamed,
Thanks for your interest!
The current code within NetworkX is in the
networkx/algorithms/isomorphism/ismags.py
module
<https://github.com/networkx/networkx/blob/main/networkx/algorithms/isomorphism/ismags.py>.
That should include only a subset of the full package features.
Searching google for "github ismags" (without quotes) leads to the github
repository for ISMAGS (java version), and to the SandiaLabs
ISMAGS-In-Python repository.

That is probably too much info to start with :)   So take an overview look
rather than a deep dive. My understanding is that Sandia Labs has written a
python library that provides the features of the java library. This is
probably a different API from the one we have chosen -- but not too
different I suspect. And they have implemented the entire library while we
did not.

Enjoy!
Dan

**Reply by elshorbagyx** [reply]
Thank you, @dschult, for your response.

I have reviewed the ISMAGS paper referenced in the documentation. Additionally, I have compared implementations between Sandia Labs and NetworkX. The process of incorporation is somewhat challenging to detail in the proposal, as both implementations require time to quantify which is better in terms of coverage, performance, and space efficiency.

My question is as follows: to what extent should the **schedule of deliverables** be detailed? Also, are there any hints or directions I can use to better formulate a timeline?

Thank you.

**Comment by dschult** [comment]
I think it is important to make the schedule of deliverables general enough to allow for details to be incorrect and the overall schedule to still be correct. For example, you shouldn't state a schedule for when each feature will be implemented. Instead, talk about groups of features. You could also add phrases like "features that have obvious implementations in NX" which lets you focus on what is doable, and then come back to the less obvious cases later. Remember to leave time for documentation and tests.

Writing a schedule of deliverables inherently involves choosing levels of detail. I'm not sure how to specify the extent it should be detailed. Too much and it becomes unrealistic. Too little and there is no sense of direction. So you are looking for a roadmap you could follow across the summer with enough freedom to adjust when things don't work the way you planned.

**Comment by elshorbagyx** [comment]
Hello @dschult,

Thank you for your response. I have written a draft of the proposal, and here it is:
https://github.com/mohamedrezk122/gsoc-24-networkx-proposal/blob/main/proposal-draft.pdf

I would appreciate it if you could give me a quick review regarding the plan and the content, and if there is something the should be edited, removed, or added.

**Comment by elshorbagyx** [comment]
Hi @dschult , 
The deadline is approaching, so if you could provide me with a quick feedback, I would be very grateful.

Thanks 😊

**Comment by dschult** [comment]
The proposal looks good. I guess that the references should say where those documents can be found. But this isn't a paper -- it's a proposal, so that's not a big deal. I don't have any more suggestions for you. :)

**Reply by elshorbagyx** [reply]
Thanks @dschult, I really appreciate your feedback.  
The references are in the form of hyperlinks. I guess they would look better with bibtex.
